src/gui/widgets/LinkedModelGroupLayout.cpp

```diff
@@ -0,0 +1,303 @@
+**     contributors may be used to endorse or promote products derived
+**     from this software without specific prior written permission.
+**
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+** "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+** LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+** A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+** OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+** LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
+**
+** $QT_END_LICENSE$
+**
+****************************************************************************/
+
+
+#include "LinkedModelGroupLayout.h"
+
+#include <QWidget>
+#include <QLayoutItem>
+#include <QLineEdit>
+#include <QRect>
+#include <QString>
+
+// TODO: rename: ModelGroupLayout
+
+constexpr const int LinkedModelGroupLayout::m_minWidth;
+
+LinkedModelGroupLayout::LinkedModelGroupLayout(QWidget *parent, int margin, int hSpacing, int vSpacing)
+	: QLayout(parent), m_hSpace(hSpacing), m_vSpace(vSpacing),
+	  m_searchBar(new QLineEdit(parent))
+{
+	setContentsMargins(margin, margin, margin, margin);
+	m_searchBar->setPlaceholderText("filter");
+	connect(m_searchBar, SIGNAL(textChanged(const QString&)),
+		this, SLOT(onTextChanged(const QString& )));
+	addWidget(m_searchBar);
+	m_searchBar->setHidden(true); // nothing to filter yet
+}
+
+LinkedModelGroupLayout::~LinkedModelGroupLayout()
+{
+	QLayoutItem *item;
+	while ((item = takeAt(0))) { delete item; }
+}
+
+void LinkedModelGroupLayout::onTextChanged(const QString&)
+{
+	invalidate();
+	update();
+}
+
+void LinkedModelGroupLayout::addItem(QLayoutItem *item)
+{
+	QWidget* widget = item->widget();
+	const QString str = widget ? widget->objectName() : QString("unnamed");
+	m_itemMap.insert(str, item);
+	invalidate();
+}
+
+int LinkedModelGroupLayout::horizontalSpacing() const
+{
+	if (m_hSpace >= 0) { return m_hSpace; }
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutHorizontalSpacing);
+	}
+}
+
+int LinkedModelGroupLayout::verticalSpacing() const
+{
+	if (m_vSpace >= 0) { return m_vSpace; }
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutVerticalSpacing);
+	}
+}
+
+int LinkedModelGroupLayout::count() const
+{
+	return m_itemMap.size() - 1;
+}
+
+QMap<QString, QLayoutItem*>::const_iterator
+LinkedModelGroupLayout::pairAt(int index) const
+{
+	if (index < 0)
+		return m_itemMap.cend();
+	QMap<QString, QLayoutItem*>::const_iterator itr = m_itemMap.cbegin();
+	++itr; // skip search bar
+	while (index-->0 && itr != m_itemMap.cend()) { ++itr; }
```

Comment:
```suggestion
	while (index-- > 0 && itr != m_itemMap.cend()) { ++itr; }
```

Reply:
I used the arrow operator on purpose to make it more clear.

Reply:
Arrow operator is `->`, not `-->`. Did you mean this to be `index->0` then?

Reply:
Heh no, it's the [long arrow operator](https://stackoverflow.com/questions/1642028/what-is-the-operator-in-c). Maybe I should write it as `index --> 0` to make it more clear?

Reply:
Okay even that just says it's two operators, specifically `--` and `>`, and because of order of operators it just happens to look like that. Use `index-- > 0`.

Reply:
OK, I'll fix. `clang-format` would have fixed it anyways.